### PR TITLE
Upgrade to Java 17.

### PR DESCRIPTION
**Pull request by cmnbroad**
Upgrades to Java 17 and gradle 7.5.1. Changes the docker to use eclipse-temurin:17-jdk as a base image (??). Uses https://github.com/openjdk/nashorn (GPL-2.0 license) as a temporary replacement for the deprecated builtin Nashorn script engine until we settle on a permanent replacement.

Als fixes https://github.com/broadinstitute/picard/issues/1840.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -68,40 +60,37 @@ ensureBuildPrerequisites(buildPrerequisitesMessage)
...
+    implementation 'org.broadinstitute:barclay:4.0.2'
+    implementation 'org.apache.logging.log4j:log4j-api:2.17.1'
+    implementation 'org.apache.logging.log4j:log4j-core:2.17.1'
+
+    // replacement for jdk nashorn engine GPL-2.0 license
+    implementation 'org.openjdk.nashorn:nashorn-core:15.4'
+
+    //NOTE: this dependency spec causes doc gen to fail on FingerPrintChecker since doc gen is not
+    // a compile time task
     compileOnly(googleNio)
```
Comment (lbergelson) [inline build.gradle:90]: Do we know what configuraton javadoc uses by default?  I'm sure we can set the javadoc task to use the compile configuration if necessary.

File: build.gradle [hunk 2]
```diff
@@ -111,8 +100,14 @@ configurations.all {
     }
 }
 
-sourceCompatibility = 1.8
-targetCompatibility = 1.8
+sourceCompatibility = 1.17
```
Comment (lbergelson) [inline build.gradle:103]: We should think about moving to toolchain definitions.

File: build.gradle [hunk 3]
```diff
@@ -10,7 +8,7 @@ buildscript {
 
 plugins {
     id "java"
```
Comment (lbergelson) [inline build.gradle:10]: We might want to migrate to java-library instead of plain java so we support the API configuration.

File: build.gradle [hunk 4]
```diff
@@ -111,8 +100,14 @@ configurations.all {
...
 }
 
-sourceCompatibility = 1.8
-targetCompatibility = 1.8
+sourceCompatibility = 1.17
+targetCompatibility = 1.17
+
+run {
+    jvmArgs = [
+            '--add-opens', 'java.base/java.io=ALL-UNNAMED'
```
Comment (lbergelson) [inline build.gradle:108]: What in picard is requiring this?   Is there any way we can avoid it?  I feel like this requirement is going to cause a ton of confusion for picard users.

Comment (cmnbroad) [reply build.gradle:108]: MergeBamAlignmentTest - I've changed it now so its only for test tasks.

File: README.md [hunk 5]
```diff
@@ -17,8 +17,7 @@ Picard is implemented using the HTSJDK Java library [HTSJDK][1] to support
 accessing file formats that are commonly used for high-throughput
 sequencing data such as [SAM][2] and [VCF][3].  
 
-Picard now builds and passes tests under Java 11. This should be considered to be a *Beta* feature. 
-As of version 2.0.1 (Nov. 2015) Picard requires Java 1.8 (jdk8u66). The last version to support Java 1.7 was release 1.141.
+As of version 3.0 (Nov. 2022) Picard requires Java 1.17.
```
Comment (lbergelson) [inline README.md:20]: We need to remember to bump this.

Comment (cmnbroad) [reply README.md:19]: Wow, that is way off.

File: build.gradle [hunk 6]
```diff
@@ -325,48 +324,63 @@ signing {
...
-                        name 'Picard Team'
-                        url 'http://broadinstitute.github.io/picard'
-                    }
-                }
+publishing {
+    publications {
+        picard(MavenPublication) {
+            from components.java
+
+            artifact sourcesJar
```
Comment (lbergelson) [inline build.gradle:332]: Future note when updating gradle build on its own.
We probably want to consider using the newer

```
java {
   withSourcesJar()
   withJavadocJar()
}
```
That handles attaching them to the publishing configuration automatically.

File: build.gradle [hunk 7]
```diff
@@ -325,48 +324,63 @@ signing {
...
     }
+    repositories {
+        maven {
+            url = "https://oss.sonatype.org/service/local/staging/deploy/maven2/"
+            credentials {
+                username = project.findProperty("sonatypeUsername")
+                password = project.findProperty("sonatypePassword")
+            }
+        }
+        maven {
```
Comment (lbergelson) [inline build.gradle:371]: Does this work?   In gatk / htsjdk we do this:

```

    repositories {
        maven {
            credentials {
                username = isRelease ? project.findProperty("sonatypeUsername") : System.env.ARTIFACTORY_USERNAME
                password = isRelease ? project.findProperty("sonatypePassword") : System.env.ARTIFACTORY_PASSWORD
            }
            def release = "https://oss.sonatype.org/service/local/staging/deploy/maven2/"
            def snapshot = "https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot-local/"
            url = isRelease ? release : snapshot
        }
    }
```

File: src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        final List<String> docArgList = new ArrayList<>();
+        docArgList.addAll(Arrays.asList(argArray));
+        docArgList.add("-cp");
+        docArgList.add(System.getProperty("java.class.path"));
+        docArgList.addAll(Arrays.asList(docTestPackages));
+
+        // Run javadoc in the current JVM with the custom doclet, and make sure it succeeds (this is a smoke test;
+        // we just want to make sure it doesn't blow up).
+        ToolProvider jdProvider = null;
```
Comment (lbergelson) [inline src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java:57]: Would it be cleaner to use ToolProvider.findFirst("javadoc")?
```
ToolProvider jdProvider = ToolProvider.findFirst("javadoc")
                .orElseThrow(() ->  new IllegalStateException("Can't find javadoc tool"));
``` ?

File: src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.spi.ToolProvider;
+
+/**
+ * Smoke test to run doc gen on a subset of classes to make sure it doesn't regress.
+ */
+public class DocumentationGenerationIntegrationTest {
+    /**
+     * Entry point for manually running the picardDoc process on a subset of packages from within picard.
+     */
+    private static String[] docTestPackages = {
```
Comment (lbergelson) [inline src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java:25]: We could probably fix this issue by re-arranging some things but lets not do it now.

File: src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            // note that the fingerprint package can't be in this list, since it requires the google
+            // cloud nio library to be present to compile
+            "picard.analysis",
+            "picard.metrics"
+    };
+
+    // suppress deprecation warning on Java 11 since we're using deprecated javadoc APIs
```
Comment (lbergelson) [inline src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java:32]: This note is out of date now?

File: src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        final String[] args = docArgList.toArray(new String[] {});
+        final int retCode = jdProvider.run(System.out, System.err, args);
+
+        // make sure the task succeeded, and generated at least one index file, plus some other files
+        Assert. assertEquals(retCode, 0);
+
+        final File outputDir = docTestTarget.toFile();
+        final Set pathNames = new HashSet<>(Arrays.stream(outputDir.list()).toList());
```
Comment (lbergelson) [inline src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java:75]: `Set -> Set<String>`

**Review by lbergelson — APPROVED** [review]
@cmnbroad A few minor issues plus a lot of rambling about how we can improve the gradle file in the future.  Good to merge when the comments are resolved!

File: .github/workflows/tests.yml [hunk 12]
```diff
@@ -12,12 +12,9 @@ jobs:
...
     strategy:
       matrix:
-        java: [ 8, 11 ]
+        java: [ 17 ]
         experimental: [false]
         run_barclay_tests: [true, false]
-#        include:
-#          - java: 17
-#            experimental: true
       fail-fast: false
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply .github/workflows/tests.yml:18]: done.

File: build.gradle (hunk 7 above)
Comment (cmnbroad) [reply build.gradle:371]: I'll give it a test and see what happens.

File: src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java (hunk 8 above)
Comment (cmnbroad) [reply src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java:57]: Yeah, better. I need to remember to do this in GATK as well.

File: src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java (hunk 10 above)
Comment (cmnbroad) [reply src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java:32]: Done.

File: src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java (hunk 11 above)
Comment (cmnbroad) [reply src/test/java/picard/util/help/DocumentationGenerationIntegrationTest.java:75]: Done.

**Review by lbergelson — APPROVED** [review]
@cmnbroad Changes look good to me.

**Pull request merged**